core/src/test/java/kafka/server/share/SharePartitionTest.java

```diff
@@ -6190,6 +6198,289 @@ public void testFindLastOffsetAcknowledgedWhenGapAtBeginning() {
+            TimestampType.CREATE_TIME, 0, 2)) {
+            // Append only 2 records for 0 offset batch starting from offset 1.
+            memoryRecords(2, 1).records().forEach(builder::append);
+        }
+        // Do not include batch from offset 5. And compact batch starting at offset 20.
+        try (MemoryRecordsBuilder builder = MemoryRecords.builder(buffer, Compression.NONE,
+            TimestampType.CREATE_TIME, 20, 2)) {
+            // Append 2 records for 20 offset batch starting from offset 20.
+            memoryRecords(2, 20).records().forEach(builder::append);
+            // And append 2 records matching the end offset of the batch.
+            memoryRecords(2, 33).records().forEach(builder::append);
+        }
+        // Send the full batch at offset 40.
+        memoryRecordsBuilder(buffer, 5, 40).close();
+        // Do not include batch from offset 45. And compact the batch at offset 50.
+        try (MemoryRecordsBuilder builder = MemoryRecords.builder(buffer, Compression.NONE,
+            TimestampType.CREATE_TIME, 50, 2)) {
+            // Append 5 records for 50 offset batch starting from offset 51.
+            memoryRecords(5, 51).records().forEach(builder::append);
+            // Append 2 records for in middle of the batch.
+            memoryRecords(2, 58).records().forEach(builder::append);
+            // And append 1 record prior to the end offset.
+            memoryRecords(1, 63).records().forEach(builder::append);
+        }
+        buffer.flip();
+        records = MemoryRecords.readableRecords(buffer);
+        // Acquire the new compacted batches. The acquire method determines the acquisition range using
+        // the first and last offsets of the fetched batches and acquires all available cached batches
+        // within that range. That means the batch from offset 45-49 which is not included in the
+        // fetch response will also be acquired. Similarly, for the batch from offset 5-19 which is
+        // anyway in the bigger cached batch of 0-34, will also be acquired. This avoids iterating
+        // through individual fetched batch boundaries; the client is responsible for reporting any
+        // data gaps via acknowledgements. This test also covers the edge case where the last fetched
+        // batch is compacted, and its last offset is before the previously cached version's last offset.
+        // In this situation, the last batch's offset state tracking is initialized. This is handled
+        // correctly because the client will send individual offset acknowledgements, which require offset
+        // state tracking anyway. While this last scenario is unlikely in practice (as a batch's reported
+        // last offset should remain correct even after compaction), the test verifies its proper handling.
+        fetchAcquiredRecords(sharePartition, records, 59);
+        assertEquals(64, sharePartition.nextFetchOffset());
```

Comment:
I think it's worth asserting the cached state here too.

Reply:
Done.